### PR TITLE
Reconstruct SELinux  mount label

### PR DESCRIPTION
**Pull request by jsafrane**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:

Reconstruct SELinux mount option `-o context=XYZ` and add it to volume manager's Actual State of World (ASW) when kubelet starts. Kubelet then can see if a mounted volume in ASW has the right context that new Pods need and can unmount it and mount back to get the right SELinux context.

SELinux context for CSI volumes is determined from json file, for in-tree volumes it's taken directly from the mount table.

~In addition, there is a new e2e test that checks the reconstruction gets the right SELinux context and volume from an old pod is un-mounted when needed.~ E2e tests will be a separate PR.

#### Does this PR introduce a user-facing change?

```release-note
Added reconstruction of SELinux mount context after kubelet restart. Feature SELinuxMountReadWriteOncePod is now fully implemented and kubelet does not lose its cache of SELinux contexts after kubelet process restart.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
https://github.com/kubernetes/enhancements/pull/3548

```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1710-selinux-relabeling
```

### CODE REVIEW

**Comment by jsafrane** [comment]
The e2e test needs some work, we can merge it after feature freeze.

File: pkg/kubelet/volumemanager/reconciler/reconstruct_new.go [hunk 1]
```diff
@@ -139,7 +142,7 @@ func (rc *reconciler) updateStatesNew(reconstructedVolumes map[v1.UniqueVolumeNa
 				klog.ErrorS(err, "Could not find device mount path for volume", "volumeName", gvl.volumeName)
 				continue
 			}
-			err = rc.actualStateOfWorld.MarkDeviceAsUncertain(gvl.volumeName, gvl.devicePath, deviceMountPath, "")
+			err = rc.actualStateOfWorld.MarkDeviceAsUncertain(gvl.volumeName, gvl.devicePath, deviceMountPath, seLinuxMountContext)
 			if err != nil {
```
Comment (gnufied) [inline pkg/kubelet/volumemanager/reconciler/reconstruct_new.go:146]: Should we have tests that verify if selinux information is recorded as expected?

File: pkg/volume/csi/csi_plugin.go [hunk 2]
```diff
@@ -455,22 +455,22 @@ func (p *csiPlugin) ConstructVolumeSpec(volumeName, mountPath string) (volume.Re
 	if err != nil {
 		return volume.ReconstructedVolume{}, errors.New(log("plugin.ConstructVolumeSpec failed loading volume data using [%s]: %v", mountPath, err))
 	}
-
 	klog.V(4).Info(log("plugin.ConstructVolumeSpec extracted [%#v]", volData))
 
-	var spec *volume.Spec
+	ret := volume.ReconstructedVolume{
+		SELinuxMountContext: volData[volDataKey.seLinuxMountContext],
```
Comment (gnufied) [inline pkg/volume/csi/csi_plugin.go:461]: Should this be feature gated as others?

**Comment by jsafrane** [comment]
Rebased, I'll address the feature gates + unit tests tomorrow.

File: pkg/kubelet/volumemanager/reconciler/reconstruct.go [hunk 3]
```diff
@@ -183,6 +183,7 @@ func (rc *reconciler) markVolumeState(volume *reconstructedVolume, volumeState o
 		VolumeGidVolume:     volume.volumeGidValue,
 		VolumeSpec:          volume.volumeSpec,
 		VolumeMountState:    volumeState,
+		SELinuxMountContext: volume.seLinuxMountContext,
 	}
```
[the remark this answers is not in the document]
Comment (jsafrane) [reply pkg/kubelet/volumemanager/reconciler/reconstruct.go:187]: It does not, removed.

File: pkg/volume/csi/csi_plugin.go (hunk 2 above)
Comment (jsafrane) [reply pkg/volume/csi/csi_plugin.go:461]: Feature gate added

File: pkg/kubelet/volumemanager/reconciler/reconstruct_new.go (hunk 1 above)
Comment (jsafrane) [reply pkg/kubelet/volumemanager/reconciler/reconstruct_new.go:146]: Unit test added

File: pkg/kubelet/volumemanager/reconciler/reconstruct_new.go [hunk 4]
```diff
@@ -139,7 +142,7 @@ func (rc *reconciler) updateStatesNew(reconstructedVolumes map[v1.UniqueVolumeNa
...
 				klog.ErrorS(err, "Could not mark device is uncertain to actual state of world", "volumeName", gvl.volumeName, "deviceMountPath", deviceMountPath)
 				continue
```
Comment (gnufied) [inline pkg/kubelet/volumemanager/reconciler/reconstruct_new.go:148]: I am trying to verify(and I forgot some details), but is staging (ie. `MountDevice`) and Setup create different json files or they create same files? Are we consistently saving selinux context in both places? Is that feature gated?

Comment (jsafrane) [reply pkg/kubelet/volumemanager/reconciler/reconstruct_new.go:148]: They create different jsons and I think SELinux is only in the SetUp one.

Comment (jsafrane) [reply pkg/kubelet/volumemanager/reconciler/reconstruct_new.go:148]: Correction: it was in MountDevice, I added it to both + both feature-gated.

**Pull request merged**